File: app/src/main/java/ru/dmitrybochkov/rates/data/repository/DefaultRepository.java
package ru.dmitrybochkov.rates.data.repository;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import androidx.lifecycle.LiveData;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Response;
import ru.dmitrybochkov.rates.data.api.v1.NetworkService;
import ru.dmitrybochkov.rates.data.db.RateDao;
import ru.dmitrybochkov.rates.domain.Rate;

/**
 * Created by Dmitry Bochkov on 14.11.2019.
 */

public class DefaultRepository implements Repository {

    private final RateDao rateDao;
    private final NetworkService networkService;

    public DefaultRepository(RateDao rateDao, NetworkService networkService) {
        this.rateDao = rateDao;
        this.networkService = networkService;
    }

    @Override
    public Flowable<Response<JsonObject>> updateRates(String base) {
        return networkService.getRates(base)
                .subscribeOn(Schedulers.io())
                .doOnNext(response -> {
                    if (response.body() != null) {
                       saveRates(response.body().getAsJsonObject("rates"));
                    }
                });

    }

    @Override
    public LiveData<List<Rate>> getRates() {
        return rateDao.loadAll();
    }

    private void saveRates(JsonObject rates) {
        Set<Map.Entry<String, JsonElement>> entries = rates.entrySet();
        ArrayList<Rate> currencies = new ArrayList<>(entries.size());
        for (Map.Entry<String, JsonElement> entry: entries) {
            currencies.add(new Rate(entry.getKey(), entry.getValue().getAsFloat()));
        }
        rateDao.insertOrReplace(currencies);
    }
}

File: app/src/main/java/ru/dmitrybochkov/rates/data/repository/DefaultRepository.kt
package ru.dmitrybochkov.rates.data.repository

import androidx.lifecycle.LiveData
import com.google.gson.JsonObject
import io.reactivex.Flowable
import io.reactivex.schedulers.Schedulers
import retrofit2.Response
import ru.dmitrybochkov.rates.data.api.v1.NetworkService
import ru.dmitrybochkov.rates.data.db.RateDao
import ru.dmitrybochkov.rates.domain.Rate
import java.util.*

/**
 * Created by Dmitry Bochkov on 14.11.2019.
 */

class DefaultRepository(private val rateDao: RateDao, private val networkService: NetworkService) : Repository {

    override val rates: LiveData<List<Rate>>
        get() = rateDao.loadAll()

    override fun updateRates(base: String): Flowable<Response<JsonObject>> {
        return networkService.getRates(base)
                .subscribeOn(Schedulers.io())
                .doOnNext { response ->
                    if (response.body() != null) {
                        saveRates(response.body()!!.getAsJsonObject("rates"))
                    }
                }

    }

    private fun saveRates(rates: JsonObject) {
        val entries = rates.entrySet()
        val currencies = ArrayList<Rate>(entries.size)
        for ((key, value) in entries) {
            currencies.add(Rate(key, value.asFloat))
        }
        rateDao.insertOrReplace(currencies)
    }
}
